### PR TITLE
BCDA-4260: Update our Github actions to leverage Actions Cache

### PR DESCRIPTION
**Pull request by adamayd**
### Fixes [BCDA-4260](https://jira.cms.gov/browse/BCDA-4260)

Right now our build process takes 25 minutes using Github actions this includes build + test stages. We want to improve this process by leveraging caches.  Ultimately the cache in GH Actions has been enabled and may provide some performance improvements, but a deeper conversation about how the container image builds and runs will be able to fully utilize the cache.

### Proposed Changes

* Leverage caches to improve build and test performance using the strategy listed above

### Change Details

* Created a Gitihub Actions cache for the go build cache and dependencies directory.
* Created mounted volumes in docker-compose to make those directories available, although not accessible during builds currently.

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

Unfortunately until the docker files conversation happens, there is no available performance timing to show.  Initially a drop of 2 - 3 mins each on build and test were recorded.  This was only short lived as local testing did not take into account the cache that would persist after a build when running back to back builds.  This cache would not exist on Github actions since each build is essentially a separate instance. As of current though, you can see the cache being built, saved, and retrieved on Github at least verifying the cache is acting as supposed.

Screenshot from 2021-03-01 15-07-16

### Feedback Requested

Ideas, ideas, ideas on how to make this better.

### CODE REVIEW

**Comment by mtrang1263** [comment]
Another verification on the presence of the cache can be seen in the test stage:

Without cache:
```
Creating bcda-app_tests_run ... 
Creating bcda-app_tests_run ... done
Running unit tests and placing results/coverage in test_results/2021-03-01_23-26-56 on host...
go: downloading github.com/huandu/go-sqlbuilder v1.10.0
go: downloading github.com/google/fhir/go v0.0.0-20210120234235-b7cfb32dc82f
...
go: downloading github.com/DATA-DOG/go-sqlmock v1.5.0
∅  bcda
✓  bcda/alr/gen (1.191s) (coverage: 90.5% of statements)
✓  bcda/alr/csv (130ms) (coverage: 97.5% of statements)
✓  bcda/api (3.493s) (coverage: 66.9% of statements)
✓  bcda/api/v2 (3.477s) (coverage: 70.8% of statements)
✓  bcda/api/v1 (7.711s) (coverage: 78.9% of statements)
✓  bcda/auth/client (2.637s) (coverage: 30.1% of statements)
✓  bcda/auth/rsautils (54ms) (coverage: 78.0% of statements)
```

With cache (no dependency downloads):
```
Creating bcda-app_tests_run ... 
Creating bcda-app_tests_run ... done
Running unit tests and placing results/coverage in test_results/2021-03-01_20-23-07 on host...
∅  bcda
✓  bcda/api (3.545s) (coverage: 66.9% of statements)
✓  bcda/api/v1 (8.188s) (coverage: 78.9% of statements)
✓  bcda/api/v2 (3.649s) (coverage: 70.8% of statements)
✓  bcda/auth/client (2.794s) (coverage: 30.1% of statements)
✓  bcda/auth/rsautils (43ms) (coverage: 78.0% of statements)
```

File: docker-compose.test.yml [hunk 1]
```diff
@@ -12,6 +12,8 @@ services:
       # Remaining environment variables defined in ./shared_files/decrypted/local.env
     volumes:
       - ./test_results:/go/src/github.com/CMSgov/bcda-app/test_results
+      - ${GOPATH}/.cache:${HOME}/.cache
```
[the remark this answers is not in the document]
Comment (adamayd) [reply docker-compose.test.yml:15]: Definitely!!  Removed the env variables and volume mount from the `Makefile` and incorporated them into the `docker-compose-test.yml` file. Running them locally produced the same results locally ~ 30ish seconds.

Screenshot from 2021-03-02 09-44-47

**Review by mtrang1263 — COMMENTED** [review]
LGTM! For verification, can we run a couple of [Build jobs](https://bcda-ci.adhocteam.us/job/BCDA%20-%20Build%20and%20Package/)?

Since we've moved around the lint cache, want to make sure we can still satisfy the timeout value.

**Comment by adamayd** [comment]
Oh, those changes may have broken the build tests though.  I didn't think about that.  I will run a build, but they will probably need to exist in both places.  Plus I'm not sure of the directory structure on the jenkins build instances.

**Review by mtrang1263 — APPROVED** [review]
LGTM!

Verified on Jenkins that the lint cache is working as expected.

```
level=info msg="[config_reader] Config search paths: [./ /go/src/github.com/CMSgov/bcda-app /go/src/github.com/CMSgov /go/src/github.com /go/src /go / /root]"

level=info msg="[lintersdb] Active 10 linters: [deadcode errcheck gosimple govet ineffassign staticcheck structcheck typecheck unused varcheck]"

level=info msg="[loader] Go packages loading at mode 575 (deps|name|compiled_files|exports_file|files|imports|types_sizes) took 23.321829948s"

level=info msg="[runner/filename_unadjuster] Pre-built 0 adjustments in 48.427867ms"

level=info msg="[linters context/goanalysis] analyzers took 0s with no stages"

level=info msg="[linters context/goanalysis] analyzers took 0s with no stages"

level=info msg="[runner] Issues before processing: 365, after processing: 0"

level=info msg="[runner] Processors filtering stat (out/in): filename_unadjuster: 365/365, nolint: 0/1, cgo: 365/365, exclude-rules: 1/127, exclude: 127/365, path_prettifier: 365/365, skip_dirs: 365/365, identifier_marker: 365/365, skip_files: 365/365, autogenerated_exclude: 365/365"

level=info msg="[runner] processing took 26.571243ms with stages: identifier_marker: 16.080999ms, path_prettifier: 4.454039ms, autogenerated_exclude: 3.20993ms, exclude: 961.272µs, skip_dirs: 838.193µs, nolint: 561.636µs, exclude-rules: 356.672µs, cgo: 73.584µs, filename_unadjuster: 29.344µs, max_same_issues: 1.746µs, source_code: 647ns, diff: 592ns, uniq_by_line: 579ns, skip_files: 401ns, max_from_linter: 398ns, path_shortener: 267ns, severity-rules: 251ns, sort_results: 250ns, max_per_file_from_linter: 245ns, path_prefixer: 198ns"

level=info msg="[runner] linters took 703.370599ms with stages: goanalysis_metalinter: 598.995223ms, unused: 77.666535ms"

level=info msg="File cache stats: 0 entries of total size 0B"

level=info msg="Memory: 243 samples, avg is 71.6MB, max is 72.7MB"

level=info msg="Execution took 24.18960806s"
```

**Pull request merged**